backend/de.metas.async/src/main/java/de/metas/async/processor/impl/planner/QueueProcessorPlanner.java

```diff
@@ -0,0 +1,360 @@
+				return;
+			}
+
+			this.startPlanner();
+		}
+		finally
+		{
+			mainLock.unlock();
+		}
+	}
+
+	public void shutdown()
+	{
+		mainLock.lock();
+		try
+		{
+			if (!queueProcessors.isEmpty())
+			{
+				queueProcessors
+						.values()
+						.stream()
+						.map(IQueueProcessor::getQueueProcessorId)
+						.forEach(this::removeQueueProcessor);
+			}
+
+			this.shutdownPlanner();
+		}
+		finally
+		{
+			mainLock.unlock();
+		}
+	}
+
+	private boolean run0()
+	{
+		final List<IQueueProcessor> availableProcessors = queueProcessors
+				.values()
+				.stream()
+				.filter(IQueueProcessor::isAvailableToWork)
+				.collect(ImmutableList.toImmutableList());
+
+		if (availableProcessors.isEmpty())
+		{
+			logger.debug("run0 - all processors are busy! Skipping...");
+			return false;
+		}
+
+		final Properties workPackageCtx = Env.newTemporaryCtx();
+
+		final List<I_C_Queue_WorkPackage> workPackages = pollAndLockWorkPackages(workPackageCtx, availableProcessors);
+		if (workPackages.isEmpty())
+		{
+			logger.debug("pollAndLockWorkpackages - returned no workPackage to be processed for availableProcessorIds={}", availableProcessors);
+			return false;
+		}
+
+		return handleWorkPackages(workPackageCtx, workPackages);
+	}
+
+	private boolean handleWorkPackages(@NonNull final Properties ctx, @NonNull final List<I_C_Queue_WorkPackage> workPackages)
+	{
+		if (workPackages.isEmpty())
+		{
+			return false;
+		}
+
+		boolean handledAllWorkPackages = true;
+
+		for (final I_C_Queue_WorkPackage workPackage : workPackages)
+		{
+			if (!isValid(workPackage))
+			{
+				WorkPackageLockHelper.unlockNoFail(workPackage);
+				handledAllWorkPackages = false;
+				continue;
+			}
+
+			final Optional<IQueueProcessor> queueProcessorOptional = getAvailableQueueProcessorForWorkPackage(workPackage);
+
+			if (!queueProcessorOptional.isPresent())
+			{
+				logger.warn("*** processWorkPackages: No available QueueProcessor found for workPackageId: {}, c_queue_workpackage.c_queue_packageprocessor_id: {}",
+							workPackage.getC_Queue_WorkPackage_ID(), workPackage.getC_Queue_PackageProcessor_ID());
+
+				WorkPackageLockHelper.unlockNoFail(workPackage);
+				handledAllWorkPackages = false;
+				continue;
+			}
+
+			final IQueueProcessor queueProcessor = queueProcessorOptional.get();
+
+			final Optional<I_C_Queue_WorkPackage> workPackageWithOwnCtx = getWorkPackageWithOwnContext(queueProcessor, workPackage, ctx);
+
+			if (!workPackageWithOwnCtx.isPresent())
+			{
+				logger.warn("*** processWorkPackages: Couldn't assign a new Ctx to C_Queue_WorkPackage: {}", workPackage.getC_Queue_WorkPackage_ID());
+
+				WorkPackageLockHelper.unlockNoFail(workPackage);
+				handledAllWorkPackages = false;
+				continue;
+			}
+
+			final boolean successfullyHandled = handleWorkPackageProcessing(queueProcessor, workPackageWithOwnCtx.get());
+
+			handledAllWorkPackages = handledAllWorkPackages && successfullyHandled;
```

Comment:
not sure if a package which could not be handled because there are no workers atm to handle it 
shall be considered a failure.

What's the reason behind?
Couldn't be that somebody in meantime has "stolen" the worker?

Reply:
or couldn't be that for some reason the number of retrieved workpackages was wrong because it was just an "approximation" ?